### ISSUE TITLE
/FI Compiler Option | Precompiled Headers

### ISSUE THREAD

**Issue by sccoouut** [mstorsjo/msvc-wine#41]
Hey, I have discovered some problem with [/FI Compiler Option](https://docs.microsoft.com/en-us/cpp/build/reference/fi-name-forced-include-file?view=msvc-170). When it is being used, compiler fails to build a project, while same project without it is built fine.

Steps to reproduce:
1. Get a compiliable project.
2. Add using of /FI Compiler Option (target_precompiled_headers for example in CMake).

CMake log example:
```
[5/9] Building CXX object CMake...ox_runner.dir/cmake_pch.cxx.obj
FAILED: CMakeFiles/sandbox_runner.dir/cmake_pch.cxx.obj 
/opt/msvc/bin/x64/cl  /nologo /TP  -I../service/include -I../ioctl/include /DWIN32 /D_WINDOWS /GR /EHsc /O1 /Ob1 /DNDEBUG -MT   -std:c++17 /Yc/src/build/CMakeFiles/sandbox_runner.dir/cmake_pch.hxx /Fp/src/build/CMakeFiles/sandbox_runner.dir/cmake_pch.cxx.pch /FI/src/build/CMakeFiles/sandbox_runner.dir/cmake_pch.hxx /showIncludes /FoCMakeFiles/sandbox_runner.dir/cmake_pch.cxx.obj /FdCMakeFiles/sandbox_runner.dir/ /FS -c CMakeFiles/sandbox_runner.dir/cmake_pch.cxx
cl : Command line error D8004 : '/FI' requires an argument
[6/9] Linking CXX static library service/service.lib
ninja: build stopped: subcommand failed.
```

May be someone has faced that issue and has found solution.

Thanks in advance.

**Comment by Palmr** [comment]
I have also experienced this issue, full build log here https://gist.github.com/Palmr/8f6b703aeaaf0a0169d3dcfe13580d1a

The only work around we found so far was to run `sed` over the `build.ninja` file and replace `/home/wine` paths with `Z:/home/wine`

**Comment by mstorsjo** [comment]
Can someone provide a minimal barebones cmake example that uses this option/feature, to let me reproduce the issue? (I don't think any of the cmake projects I've been using recently uses this option, or the projects are large and unsuitable for testing/debugging.)

**Comment by Palmr** [comment]
https://github.com/Palmr/msvc-wine_issue-41_repro

That's a cut down test case that reproduces the issue for me locally (I'm using https://www.jetbrains.com/clion/ and the docker toolchain: https://www.jetbrains.com/help/clion/clion-toolchains-in-docker.html)

To get around this we were mounting our code in to the container and running:

```bash
CC=cl CXX=cl cmake -G Ninja -DCMAKE_C_COMPILER_WORKS=1 -DCMAKE_BUILD_TYPE=Release -DCMAKE_SYSTEM_NAME=Windows -DCMAKE_CROSSCOMPILING=ON .

for flag in "FI" "Yc" "Yu"; do
       sed -i -r "s#/${flag}/home/wine#/${flag}Z:/home/wine#g" build.ninja
done

# see https://github.com/mstorsjo/msvc-wine/issues/6
sed -i -r 's#/Zi#/Z7#g' build.ninja

ninja
```

**Comment by mstorsjo** [comment]
> https://github.com/Palmr/msvc-wine_issue-41_repro
> 
> That's a cut down test case that reproduces the issue for me locally (I'm using https://www.jetbrains.com/clion/ and the docker toolchain: https://www.jetbrains.com/help/clion/clion-toolchains-in-docker.html)
> 
> To get around this we were mounting our code in to the container and running:
> 
> ```shell
> CC=cl CXX=cl cmake -G Ninja -DCMAKE_C_COMPILER_WORKS=1 -DCMAKE_BUILD_TYPE=Release -DCMAKE_SYSTEM_NAME=Windows -DCMAKE_CROSSCOMPILING=ON .
> 
> for flag in "FI" "Yc" "Yu"; do
>        sed -i -r "s#/${flag}/home/wine#/${flag}Z:/home/wine#g" build.ninja
> done
> 
> # see https://github.com/mstorsjo/msvc-wine/issues/6
> sed -i -r 's#/Zi#/Z7#g' build.ninja
> 
> ninja
> ```

Thanks for the repro and for the tips about how you work around it!

So using this feature in CMake essentially boils down to using `target_precompile_headers` to specify one header?

I made one attempt at a fix for the issue, see https://github.com/mstorsjo/msvc-wine/commit/abs-path. Would that fix the issue for you? (I didn't merge it into the master branch quite yet.)

**Comment by Palmr** [comment]
https://gist.github.com/Palmr/f1ba2de8e6da3a84c334dd2e6e1c8e87

It seems unhappy, though it looks like CMake is finding `cl.exe` at `Z:\opt\msvc\vc\tools\msvc\14.31.31103\bin\Hostx64\x86\cl.exe` rather than the wrapper at `/opt/msvc/bin/x86/cl.exe` which would do the argument fixing.

I tried running with env vars of `CXX=/opt/msvc/bin/x86/cl;CC=/opt/msvc/bin/x86/cl` but no change.

I must admit I am not a Wine/CMake user or Windows developer normally so there is a fair chance the problem is me.

All I could notice was `/opt/msvc/bin/x86/msvcenv.sh` (which is wrapped) seems to set:

```bash
BINDIR=$BASE_UNIX/vc/tools/msvc/$MSVCVER/bin/Hostx64/$ARCH
export WINEPATH="${BINDIR//\//\\};${SDKBINDIR//\//\\};$MSVCDIR\\bin\\Hostx64\\x64"
```

And maybe that's the `PATH` it's using to find `cl.exe`?

**Comment by mstorsjo** [comment]
> https://gist.github.com/Palmr/f1ba2de8e6da3a84c334dd2e6e1c8e87
> 
> It seems unhappy, though it looks like CMake is finding `cl.exe` at `Z:\opt\msvc\vc\tools\msvc\14.31.31103\bin\Hostx64\x86\cl.exe` rather than the wrapper at `/opt/msvc/bin/x86/cl.exe` which would do the argument fixing.

Did this work before you applied the change? Because that seems quite unrelated to this patch.

> All I could notice was `/opt/msvc/bin/x86/msvcenv.sh` (which is wrapped) seems to set:
> 
> ```shell
> BINDIR=$BASE_UNIX/vc/tools/msvc/$MSVCVER/bin/Hostx64/$ARCH
> export WINEPATH="${BINDIR//\//\\};${SDKBINDIR//\//\\};$MSVCDIR\\bin\\Hostx64\\x64"
> ```
> 
> And maybe that's the `PATH` it's using to find `cl.exe`?

That path is set up so that wine can execute `cl.exe` from there, but this directory shouldn't be in `$PATH` so that cmake would be able to pick it up.

**Comment by Palmr** [comment]
I think I read the error message incorrectly: https://gist.github.com/Palmr/f1ba2de8e6da3a84c334dd2e6e1c8e87#file-msvc-wine_issue14-attempt2-txt-L7

Line 7 does reference the correct wrapper script ` /opt/msvc/bin/x86/cl` and seems it tried to run:

```bash
/opt/msvc/bin/x86/cl @/tmp/misc/response-file12954642806412015051 /tmp/misc/compiler-file7185106763504052988 | @response-file12954642806412015051=/TP /DIssueRepro_EXPORTS /DNDEBUG /DWIN32 /D_MBCS /D_USRDLL /D_WINDOWS /EHsc /EHsc -MD /GL /Oi /GF /Gy /W3 /Zi /EHsc /FI/tmp/msvc-wine_issue-41_repro/cmake-build-release-docker-wine/IssueRepro/CMakeFiles/IssueRepro.dir/cmake_pch.hxx /Be /Bd /EP
```

---

I modified the patch to dump all arguments it sees to `stderr` and exiting (code 123) before actually invoking the wrapper executable and get the following:

```
Cannot get compiler information:
	Compiler exited with error code 123: /opt/msvc/bin/x86/cl @/tmp/misc/response-file35818133934357952 /tmp/misc/compiler-file4849077582937730181 | @response-file35818133934357952=/TP /DIssueRepro_EXPORTS /DNDEBUG /DWIN32 /D_MBCS /D_USRDLL /D_WINDOWS /EHsc /EHsc -MD /GL /Oi /GF /Gy /W3 /Zi /EHsc /Be /Bd /EP
	A= @/tmp/misc/response-file35818133934357952
	A= /tmp/misc/compiler-file4849077582937730181
	
	
	Compiler exited with error code 123: /opt/msvc/bin/x86/cl @/tmp/misc/response-file10158653504850592337 /tmp/misc/compiler-file4849077582937730181 | @response-file10158653504850592337=/TP /DIssueRepro_EXPORTS /DNDEBUG /DWIN32 /D_MBCS /D_USRDLL /D_WINDOWS /EHsc /EHsc -MD /GL /Oi /GF /Gy /W3 /Zi /EHsc /FI/tmp/msvc-wine_issue-41_repro/cmake-build-release-docker-wine/IssueRepro/CMakeFiles/IssueRepro.dir/cmake_pch.hxx /Be /Bd /EP
	A= @/tmp/misc/response-file10158653504850592337
	A= /tmp/misc/compiler-file4849077582937730181
	
	
	Compiler exited with error code 123: /opt/msvc/bin/x86/cl @/tmp/misc/response-file9959576051178754429 /tmp/misc/compiler-file4849077582937730181 | @response-file9959576051178754429=/TP /DWIN32 /EHsc /EHsc -MD /GL /Zi /EHsc /Be /Bd /EP
	A= @/tmp/misc/response-file9959576051178754429
	A= /tmp/misc/compiler-file4849077582937730181
	
	
	Compiler exited with error code 123: /opt/msvc/bin/x86/cl @/tmp/misc/response-file13474969894823922258 /tmp/misc/compiler-file4849077582937730181 | @response-file13474969894823922258=/TP /DWIN32 /EHsc /EHsc -MD /GL /Zi /EHsc /FI/tmp/msvc-wine_issue-41_repro/cmake-build-release-docker-wine/IssueRepro/CMakeFiles/IssueReproTest.dir/cmake_pch.hxx /Be /Bd /EP
	A= @/tmp/misc/response-file13474969894823922258
	A= /tmp/misc/compiler-file4849077582937730181
```

I can see the while loop over arguments doesn't have any effect on the `/FI...` because it comes after a pipe in my case.

**Comment by mstorsjo** [comment]
> I think I read the error message incorrectly: https://gist.github.com/Palmr/f1ba2de8e6da3a84c334dd2e6e1c8e87#file-msvc-wine_issue14-attempt2-txt-L7
> 
> Line 7 does reference the correct wrapper script ` /opt/msvc/bin/x86/cl` and seems it tried to run:
> 
> ```shell
> /opt/msvc/bin/x86/cl @/tmp/misc/response-file12954642806412015051 /tmp/misc/compiler-file7185106763504052988 | @response-file12954642806412015051=/TP /DIssueRepro_EXPORTS /DNDEBUG /DWIN32 /D_MBCS /D_USRDLL /D_WINDOWS /EHsc /EHsc -MD /GL /Oi /GF /Gy /W3 /Zi /EHsc /FI/tmp/msvc-wine_issue-41_repro/cmake-build-release-docker-wine/IssueRepro/CMakeFiles/IssueRepro.dir/cmake_pch.hxx /Be /Bd /EP
> ```
> 
> I modified the patch to dump all arguments it sees to `stderr` and exiting (code 123) before actually invoking the wrapper executable and get the following:
> 
> ```
> Cannot get compiler information:
> 	Compiler exited with error code 123: /opt/msvc/bin/x86/cl @/tmp/misc/response-file35818133934357952 /tmp/misc/compiler-file4849077582937730181 | @response-file35818133934357952=/TP /DIssueRepro_EXPORTS /DNDEBUG /DWIN32 /D_MBCS /D_USRDLL /D_WINDOWS /EHsc /EHsc -MD /GL /Oi /GF /Gy /W3 /Zi /EHsc /Be /Bd /EP
> 	A= @/tmp/misc/response-file35818133934357952
> 	A= /tmp/misc/compiler-file4849077582937730181
> 	
> 	
> 	Compiler exited with error code 123: /opt/msvc/bin/x86/cl @/tmp/misc/response-file10158653504850592337 /tmp/misc/compiler-file4849077582937730181 | @response-file10158653504850592337=/TP /DIssueRepro_EXPORTS /DNDEBUG /DWIN32 /D_MBCS /D_USRDLL /D_WINDOWS /EHsc /EHsc -MD /GL /Oi /GF /Gy /W3 /Zi /EHsc /FI/tmp/msvc-wine_issue-41_repro/cmake-build-release-docker-wine/IssueRepro/CMakeFiles/IssueRepro.dir/cmake_pch.hxx /Be /Bd /EP
> 	A= @/tmp/misc/response-file10158653504850592337
> 	A= /tmp/misc/compiler-file4849077582937730181
> 	
> 	
> 	Compiler exited with error code 123: /opt/msvc/bin/x86/cl @/tmp/misc/response-file9959576051178754429 /tmp/misc/compiler-file4849077582937730181 | @response-file9959576051178754429=/TP /DWIN32 /EHsc /EHsc -MD /GL /Zi /EHsc /Be /Bd /EP
> 	A= @/tmp/misc/response-file9959576051178754429
> 	A= /tmp/misc/compiler-file4849077582937730181
> 	
> 	
> 	Compiler exited with error code 123: /opt/msvc/bin/x86/cl @/tmp/misc/response-file13474969894823922258 /tmp/misc/compiler-file4849077582937730181 | @response-file13474969894823922258=/TP /DWIN32 /EHsc /EHsc -MD /GL /Zi /EHsc /FI/tmp/msvc-wine_issue-41_repro/cmake-build-release-docker-wine/IssueRepro/CMakeFiles/IssueReproTest.dir/cmake_pch.hxx /Be /Bd /EP
> 	A= @/tmp/misc/response-file13474969894823922258
> 	A= /tmp/misc/compiler-file4849077582937730181
> ```
> 
> I can see the while loop over arguments doesn't have any effect on the `/FI...` because it comes after a pipe in my case.

So it seems like your cmake invocation decides to pass these arguments via a response file (i.e. storing the arguments in a file and passing that file to the command, instead of passing the arguments directly on the command line) for some unknown reason. For such cases, the script won't be able to fix them; the existing handling of absolute paths won't be able to handle those cases either. (I presume that your "comes after a pipe" is cmake's way of telling you what the response file contained.)

When I tried building your repro example, I didn't see any such issues, and building worked fine (except that the repro example isn't buildable as such) once I applied this fix though. So something is odd with how you invoke cmake.

**Comment by Palmr** [comment]
Ah, having not done much windows development I wasn't aware of [what the @-prefixed arguments meant](https://docs.microsoft.com/en-us/cpp/build/reference/at-specify-a-compiler-response-file?view=msvc-170).

I patched your patch with something rather crude below in the case statement and it appears to work now:

```bash
        [@]/*)
           path=${a:1}
           for flag in "FI" "Yc" "Yu"; do
                  sed -i -r "s#/${flag}/#/${flag}Z:/#g" ${path}
           done
           sed -i -r 's#/Zi#/Z7#g' ${path}
           ;;
```

Pretty sure that could be improved upon, and to cover the existing handling of absolute paths.

I'm not sure why CMake is invoking `cl` differently on my machine I'm afraid.

**Comment by Palmr** [comment]
Having slightly taken over poor @sccoouut 's issue here, feels like this is now a separate issue that I could raise for handling compiler response files and a better fix than my `sed`'s could be applied to fix that.

The fix you put on a branch originally should be enough to close this ticket.

**Comment by mstorsjo** [comment]
> Ah, having not done much windows development I wasn't aware of [what the @-prefixed arguments meant](https://docs.microsoft.com/en-us/cpp/build/reference/at-specify-a-compiler-response-file?view=msvc-170).

FWIW, response files are a thing outside of Windows too, but they're much less used there as the command line on other OSes don't have as tight limitations on length.

> The fix you put on a branch originally should be enough to close this ticket.

Ok, I pushed that fix as 4a1b93df8bbca606381c8423ce009cdd2af3c5a6 now, so the main issue here should be resolved.
